torchaudio/prototype/functional/_ray_tracing.py

```diff
@@ -0,0 +1,156 @@
+from typing import Union
+
+import torch
+
+
+def _validate_absorption_scattering(
+    v: Union[float, torch.Tensor], name: str, num_walls: int, D: int, dtype: torch.dtype
+) -> torch.Tensor:
+    """Validates and converts e_absorption or scattering parameters to a tensor with appropriate shape"""
+    if isinstance(v, float):
+        out = torch.full(
+            size=(
+                1,
+                num_walls,
+            ),
+            fill_value=v,
+        )
+    elif isinstance(v, torch.Tensor) and v.ndim == 1:
+        if v.shape[0] != num_walls:
+            raise ValueError(
+                f"The shape of {name} must be (4,) or (6,) if it is a 1D Tensor."
+                f"Found the shape of room is {D} and shape of {name} is {v.shape}."
+            )
+        out = v[None, :]
+    elif isinstance(v, torch.Tensor) and v.ndim == 2:
+        if v.shape[1] != num_walls:
+            raise ValueError(
+                f"The shape of {name} must be (num_bands, 4) for a 2D room or (num_bands, 6) "
+                "for a 3D room if it is a 2D Tensor. "
+                f"Found the shape of room is {D} and shape of {name} is {v.shape}."
+            )
+        out = v
+    else:
+        out = v
+    assert out.ndim == 2
+    out = out.to(dtype)
+
+    return out
+
+
+def ray_tracing(
+    room: torch.Tensor,
+    source: torch.Tensor,
+    mic_array: torch.Tensor,
+    num_rays: int,
+    e_absorption: Union[float, torch.Tensor] = 0.0,
+    scattering: Union[float, torch.Tensor] = 0.0,
+    mic_radius: float = 0.5,
+    sound_speed: float = 343.0,
+    energy_thres: float = 1e-7,
+    time_thres: float = 10.0,
+    hist_bin_size: float = 0.004,
+) -> torch.Tensor:
+    r"""Compute energy histogram via ray tracing.
+
+    The implementation is based on *pyroomacoustics* :cite:`scheibler2018pyroomacoustics`.
+
+    ``num_rays`` rays are casted uniformly in all directions from the source; when a ray intersects a wall,
+    it is reflected and part of its energy is absorbed. It is also scattered (sent directly to the microphone(s))
+    according to the ``scattering`` coefficient. When a ray is close to the microphone, its current energy is
+    recoreded in the output histogram for that given time slot.
+
+    .. devices:: CPU
+
+    .. properties:: TorchScript
+
+    Args:
+        room (torch.Tensor): The 1D Tensor to determine the room size. The shape is
```

Comment:
something like "1D tensor of room dimensions" or simply "room dimensions" might be clearer